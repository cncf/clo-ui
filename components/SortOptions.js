module.exports = require('../dist/cjs/components/SortOptions.js');
