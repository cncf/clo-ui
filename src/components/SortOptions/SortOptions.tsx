import { isNull, isUndefined } from 'lodash';
import React, { ChangeEvent, useRef } from 'react';

export interface ISortOptionsProps {
  width: number;
  options: any[];
  by: string;
  direction?: string;
  className?: string;
  onSortChange: (value: string) => void;
}

export const SortOptions: React.FC<ISortOptionsProps> = (props: ISortOptionsProps) => {
  const selectEl = useRef<HTMLSelectElement>(null);

  const handleChange = (event: ChangeEvent<HTMLSelectElement>) => {
    props.onSortChange(event.target.value);
    forceBlur();
  };

  const forceBlur = (): void => {
    if (!isNull(selectEl) && !isNull(selectEl.current)) {
      selectEl.current.blur();
    }
  };

  return (
    <div className={`d-flex flex-nowrap align-items-center me-2 me-md-4 ${props.className}`}>
      <label className="form-label me-2 mb-0">Sort:</label>
      <select
        ref={selectEl}
        className="form-select form-select-sm rounded-0 cursorPointer"
        style={{ width: `${props.width}px` }}
        value={`${props.by}${!isUndefined(props.direction) ? `_${props.direction}` : ''}`}
        onChange={handleChange}
        aria-label="Sort options select"
      >
        {props.options.map((opt: any) => (
          <option key={`sort_${opt.label}${!isUndefined(opt.direction) ? `_${opt.direction}` : ''}`} value={opt.by}>
            {opt.label}
          </option>
        ))}
        ;
      </select>
    </div>
  );
};
