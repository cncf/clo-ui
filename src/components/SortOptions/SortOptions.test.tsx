import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SortOptions } from './SortOptions';

const mockOnSortChange = jest.fn();

const defaultProps = {
  width: 300,
  options: [
    {
      label: 'Applied at (asc)',
      by: 'date',
      direction: 'asc',
    },
    {
      label: 'Applied at (desc)',
      by: 'date',
      direction: 'desc',
    },
  ],
  by: 'date',
  direction: 'asc',
  onSortChange: mockOnSortChange,
};

describe('SortOptions', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<SortOptions {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', () => {
    render(<SortOptions {...defaultProps} />);

    expect(screen.getByText('Sort:')).toBeInTheDocument();

    expect(screen.getByRole('combobox')).toBeInTheDocument();
    expect(screen.getAllByRole('option').length).toBe(2);
    expect(screen.getByRole('option', { name: 'Applied at (asc)' })).toBeInTheDocument();
    expect(screen.getByRole('option', { name: 'Applied at (desc)' })).toBeInTheDocument();
    expect((screen.getByRole('option', { name: 'Applied at (asc)' }) as HTMLOptionElement).selected).toBe(true);

    expect(screen.getByLabelText('Sort options select')).toBeInTheDocument();
    expect(screen.getByLabelText('Sort options select')).toHaveValue('date_asc');
  });

  it('calls onChange to update select', async () => {
    render(<SortOptions {...defaultProps} />);

    await userEvent.selectOptions(
      screen.getByRole('combobox'),
      screen.getByRole('option', { name: 'Applied at (desc)' })
    );

    expect(mockOnSortChange).toHaveBeenCalledTimes(1);
    expect(mockOnSortChange).toHaveBeenCalledWith('date_desc');
  });
});
