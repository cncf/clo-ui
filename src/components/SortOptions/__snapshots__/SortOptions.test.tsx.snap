// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SortOptions creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-nowrap align-items-center undefined"
  >
    <label
      class="form-label me-2 mb-0"
    >
      Sort:
    </label>
    <select
      aria-label="Sort options select"
      class="form-select form-select-sm rounded-0 cursorPointer"
      style="width: 300px;"
    >
      <option
        value="date_asc"
      >
        Applied at (asc)
      </option>
      <option
        value="date_desc"
      >
        Applied at (desc)
      </option>
      ;
    </select>
  </div>
</DocumentFragment>
`;
